### PR TITLE
Answer solo /help without calling the LLM

### PR DESCRIPTION
**Pull request by 2witstudios**
## Summary
- `/help` was a full billed model turn even for the common case of typing the chip alone with no other text — the command's own data (the sender's precedence-resolved command list) was already loaded deterministically, the LLM only added latency and paraphrasing risk.
- A message that is **only** the `/help` chip (no other text) now answers directly from code in all three AI entry points — page chat, the global assistant, and channel agent mentions — with no `streamText`/`generateText` call and no credit hold taken.
- `/help` combined with real text (e.g. `/help how do I use spreadsheets`) is unchanged and still goes through the model, since that's a genuine question.

## What changed
- `isSoloBuiltinCommand` (`command-processor.ts`) — pure detector for "this message is nothing but one built-in command chip." Used by the two HTTP chat routes, which don't need a mention to reach them.
- `isSoloBuiltinCommandIgnoringMentions` — same test, but tolerant of `@mention` tokens alongside the chip. Required for the channel-mention entry point: `triggerMentionedAgentResponses` only runs when the message contains at least one `@mention` (that's how it resolves which agent to reply as), so a solo `/help` there is always shaped like `@Agent /help` — under the strict variant the required mention would itself always disqualify it, making the fast path permanently unreachable. Caught this on a self-review pass; fixed with a dedicated integration test that sends realistic `@Agent /help` content through the responder and asserts it actually takes the fast path.
- `buildHelpAnswerText` (`command-core.ts`) — pure formatter for the direct, human-facing answer; sibling to `buildHelpPromptSection`, which is untouched and still serves the non-solo (LLM) path.
- `help-answer.ts` / `help-responder.ts` — load the sender's command list once and stream + persist the reply using the same `createUIMessageStream` / `buildAssistantPersistencePayload` building blocks a real model turn already uses, so it renders, streams, and survives reload identically to a normal reply. Persistence runs synchronously inside `execute` (not `onFinish`) via an injected `persist` closure — each route passes its own `saveTerminalAssistantMessage`/`saveTerminalGlobalAssistantMessage`, the same transactional, active-conversation-rechecking wrapper every other terminal assistant write uses.
- Each of the three entry points (`apps/web/src/app/api/ai/chat/route.ts`, `apps/web/src/app/api/ai/global/[id]/messages/route.ts`, `apps/web/src/lib/channels/agent-mention-responder.ts`) skips its credit gate for a solo `/help` (reusing the existing `isMeteringExempt`-style skip pattern, so no hold is ever taken/needs releasing) and returns before any provider/history/lifecycle setup runs.
- The solo-`/help` short-circuit takes the same per-conversation `startGenerationExclusive`/`takeOverConversationStreams` guard every other send does (so it correctly takes over an in-flight generation from another tab), and broadcasts the user's own `/help` message the same way a real turn does (page-chat's existing `isShared` gate; global assistant's existing unconditional-to-own-channel pattern, including the new-conversation sidebar broadcast).

## Known remaining gap
The assistant's `/help` reply itself does not get the AI-stream-start/complete websocket broadcast a real turn gets (the multiplayer live-stream protocol backing `useActiveStream`/stream-join/reconnect across ~16 client files) — replicating that for an already-complete synthetic reply risked getting the sequencing subtly wrong for shared/collaborative conversations generally. Collaborators watching a shared conversation see the `/help` chip immediately (now broadcast) but the reply itself on next refetch — a temporary staleness, not data loss, since persistence is durable. Full reasoning in the PR review thread.

## CodeQL note
CodeQL flags a `js/user-controlled-bypass` (high) on the `isSoloHelpRequest` early-return: it sees user-controlled content gating reachability of downstream provider/tool-building code. Traced via the full SARIF codeFlow: the specific "sensitive action" it points at (MCP-scope tool filtering feeding `streamText`) is code this PR never touches (byte-identical to `master`, no open alert there), and it's simply never reached on the `isSoloHelpRequest` path at all — no model runs, no tools are ever exposed, so there's nothing the skipped check was protecting that's still reachable. Believe this is a false positive; left open (not dismissed) for a second opinion given the severity classification — full trace in the PR's CodeQL review thread. Note: this alert re-registers under a new number on each push that shifts nearby lines (fingerprinting doesn't match across the shift) — same underlying finding each time, already explained on the thread.

## Test plan
- [x] `bun run --filter web typecheck` / `bun run --filter @pagespace/lib typecheck` — clean on all touched files
- [x] `bun run --filter web lint` — no new warnings
- [x] `bun run knip:check` — stays within baseline
- [x] Unit tests: `isSoloBuiltinCommand` + `isSoloBuiltinCommandIgnoringMentions` (command-processor.test.ts), `buildHelpAnswerText` (command-core.test.ts), `help-responder.test.ts` (persistence contract, pill inclusion, graceful failure) — all passing
- [x] Integration-level wiring tests in both HTTP routes' `stream-socket-events.test.ts` (takeover fires, credit gate skipped, broadcasts correct, combined `/help + text` still uses the LLM, active-conversation guard actually skips the write when the conversation goes inactive) and in `agent-mention-responder-commands.test.ts` (realistic `@Agent /help` content takes the fast path, model never called)
- [x] Existing suites unchanged and passing: `command-resolver.test.ts`, `command-resolver-edge-cases.test.ts`, `available-commands.test.ts`, `stream-takeover.test.ts`, both routes' `credit-gate.test.ts`
- [ ] Manual: send a page-chat message that is only the `/help` chip and confirm the reply is near-instant, the "Using /help" pill renders, and the message survives reload; confirm `/help how do I use spreadsheets` still goes through the model; repeat in the global assistant and a channel mention

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01Wihsv1wnwgsNb8WgNXb6hM

### CODE REVIEW

File: apps/web/src/app/api/ai/chat/route.ts [hunk 1]
```diff
@@ -928,6 +941,21 @@ export async function POST(request: Request) {
       }
     }
 
+    // The user's /help message is already durably saved above (same generic
+    // transaction as any other message) — answer it directly from code and
+    // return before any of the provider/history/lifecycle machinery below
+    // runs. No streamText, no credit hold (skipped above), nothing to stop
+    // or reconnect to since the reply is already complete.
+    if (isSoloHelpRequest) {
+      return await respondWithHelpAnswer({
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/app/api/ai/chat/route.ts:991]: Fixed — the solo-`/help` short-circuit now runs through the same `startGenerationExclusive`/`takeOverConversationStreams` every other send goes through before returning, so a solo `/help` from a second tab correctly takes over an in-flight generation on the same conversation instead of landing alongside it. See `apps/web/src/app/api/ai/chat/route.ts` around the `isSoloHelpRequest` branch (commit 82cddda54).

File: apps/web/src/lib/ai/core/help-responder.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      });
+      writer.write({ type: 'text-start', id: `${messageId}-text` });
+      writer.write({ type: 'text-delta', id: `${messageId}-text`, delta: answerText });
+      writer.write({ type: 'text-end', id: `${messageId}-text` });
+    },
+    onFinish: async ({ responseMessage }) => {
+      if (!responseMessage) return;
+      const payload = buildAssistantPersistencePayload(messageId, responseMessage.parts);
+      try {
+        await persist(payload);
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/lib/ai/core/help-responder.ts]: Fixed — persistence now runs synchronously inside `execute`, not `onFinish`. The reply's full content is known before `execute` even starts writing to the stream (no incremental generation), so there was no reason to defer the write to `onFinish` in the first place; doing it in `execute` matches the real routes'\'' own execute-end durable-write reasoning (runs regardless of client connection state). See `help-responder.ts`'\''s `buildHelpAnswerStream` (commit 82cddda54).

File: apps/web/src/lib/ai/core/help-responder.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  pageId: string;
+  conversationId: string;
+  originalMessages: UIMessage[];
+}): Promise<Response> {
+  const answerText = await loadHelpAnswerText(params.userId, params.driveId);
+  const messageId = createId();
+
+  return buildHelpAnswerStream(messageId, answerText, params.originalMessages, (payload) =>
+    saveMessageToDatabase({
+      messageId,
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/lib/ai/core/help-responder.ts]: Fixed — `help-responder.ts` no longer calls `saveMessageToDatabase`/`saveGlobalAssistantMessageToDatabase` directly. It now takes an injected `persist` closure, and each route passes its own `saveTerminalAssistantMessage`/`saveTerminalGlobalAssistantMessage` — the exact same transaction that locks the conversation row and re-checks `isActive` before writing that every other terminal assistant write in these routes goes through. A History-delete landing between the user'\''s `/help` message and this write now correctly drops the reply instead of resurrecting an orphaned active message (commit 82cddda54).

File: apps/web/src/app/api/ai/global/[id]/messages/route.ts [hunk 4]
```diff
@@ -641,6 +655,23 @@ export async function POST(
...
+    // The user's /help message is already durably saved above (same generic
+    // update as any other message) — answer it directly from code and return
+    // before any of the provider/history/lifecycle machinery below runs. No
+    // streamText, no credit hold (skipped above), nothing to stop or
+    // reconnect to since the reply is already complete. (A brand-new
+    // conversation's realtime sidebar broadcast to other tabs, which normally
+    // happens much later in this route, does not fire on this path — the
+    // conversation is still visible on next load/refetch.)
+    if (isSoloHelpRequest) {
+      return await respondWithGlobalHelpAnswer({
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/app/api/ai/global/[id]/messages/route.ts]: Partially fixed. The user'\''s own `/help` message is now broadcast the same way (and under the same gating) a real turn'\''s user message is: page-chat respects the existing `isShared` gate, the global assistant broadcasts unconditionally to the user'\''s own channel, matching each route'\''s existing pattern exactly. I also wired the previously-acknowledged new-conversation sidebar broadcast for the global assistant while I was in there, since it turned out to be nearly free once `channelId`/`displayName` were available early.

I did NOT wire the assistant reply'\''s own AI-stream-start/complete websocket broadcast (`createStreamLifecycle`'\''s `broadcastAiStreamStart`/`broadcastAiStreamComplete`, feeding `useActiveStream`/`stream-join-poll-fallback`/`planOwnStreamMirror` and similar client-side multiplayer-stream state machines across ~16 files). Deliberately: that protocol assumes a stream that starts, potentially gets joined by a reconnecting/other tab, and is heartbeat-tracked in `aiStreamSessions` — none of which exists for an already-complete synthetic reply, and getting the sequencing subtly wrong (e.g. a `stream_complete` with no matching `stream_start`/session row) risks confusing that state machine for shared/collaborative conversations generally, not just for `/help`. Given severity (P2, a temporary staleness that resolves on next refetch — no data loss, since persistence is now durable per the fix above) versus the blast radius of a subtly-wrong client protocol integration, I left this as a documented gap rather than guessing at the wiring. Collaborators watching a shared conversation live will see the `/help` chip immediately (now broadcast) but the reply itself only on next refetch. Happy to revisit with more context on the intended cross-tab-completion contract if that'\''s wanted before merge.

File: apps/web/src/app/api/ai/chat/route.ts [hunk 5]
```diff
@@ -928,6 +941,21 @@
       }
     }
 
+    // The user's /help message is already durably saved above (same generic
+    // transaction as any other message) — answer it directly from code and
+    // return before any of the provider/history/lifecycle machinery below
+    // runs. No streamText, no credit hold (skipped above), nothing to stop
+    // or reconnect to since the reply is already complete.
+    if (isSoloHelpRequest) {
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/app/api/ai/chat/route.ts:967]: I traced this fully via the SARIF codeFlow (`code-scanning/alerts/274`) rather than just the summary. Findings:

**The flagged "sensitive action"** (per the alert markdown link) is `filterToolsForMcpScope(pageSpaceTools, isScopedMCPAuth(authResult))` at (then) line 1085 — building `baseTools`, the input to the tool-filtering pipeline that feeds `streamText`. That code is **byte-identical** to `master` (currently line 1057 there — the +28 line delta is exactly the code this PR added above it). I diffed it directly: `git show master:...route.ts | sed -n "1057p"` vs this branch — same text, same AST. No open alert exists for it on `master` today (only 3 open `js/user-controlled-bypass` alerts on that file, both at lines 350/355, unrelated).

**The mechanism CodeQL is flagging**: `isSoloHelpRequest` (derived from the request body via `extractMessageContent`) gates an early `return` — reaching line 1085 (and everything after it: provider resolution, tool building, `streamText`) now requires `!isSoloHelpRequest`. CodeQL's `js/user-controlled-bypass` heuristic sees "reachability of security-relevant code X is conditioned on tainted value" and flags it, without being able to verify what I can verify by reading the branch: when `isSoloHelpRequest` is true, **no model is ever invoked and no tools are ever exposed to anything** — the MCP-scope tool-filtering result (`baseTools`) is simply never computed or used on that path, because the function returns before it's reached. There is no code reachable from the `isSoloHelpRequest === true` branch that uses tools, calls a provider, or does anything the skipped entitlement/MCP-scope checks were protecting. `isSoloBuiltinCommand` (the pure detector) also only matches a message containing *exactly* one command token and no other text — there's no room to smuggle additional content that could reach the skipped code while `isSoloHelpRequest` stays true.

So functionally: a user can trivially make `isSoloHelpRequest` true (it's just typing `/help` alone), but doing so only routes them to a free, deterministic help reply — it cannot be used to reach the tool-filtering/MCP-scope/provider code with attacker-chosen tool exposure, because that code doesn't run at all on that path. I believe this is a false positive: the "bypass" is real (the code is skipped) but the thing it protects is also entirely absent on the same path, so there's nothing to gain unauthorized access to.

I'm not dismissing this alert myself given the "high" severity classification — that's a call I think should have a second set of eyes given I authored the code in question. Posting the full trace here so whoever reviews can verify quickly (or push back if I'm missing something) and either dismiss as false-positive or ask for a restructure.

File: apps/web/src/app/api/ai/chat/route.ts [hunk 6]
```diff
@@ -928,6 +946,56 @@
...
+    // The user's /help message is already durably saved above (same generic
+    // transaction as any other message) — answer it directly from code and
+    // return before any of the provider/history/lifecycle machinery below
+    // runs. No streamText, no credit hold (skipped above). It still takes
+    // the SAME per-conversation takeover every other send does (review
+    // finding — chatgpt-codex-connector, PR #2329): without it, a solo
+    // /help sent from a second tab while another turn is generating would
+    // land alongside that turn instead of taking it over, unlike every
+    // other message.
+    if (isSoloHelpRequest) {
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/app/api/ai/chat/route.ts:967]: Same finding re-surfaced under a new alert number (`code-scanning/alerts/275`) because the line shifted (my subsequent commits added code above it) and fingerprinting didn't match it to the prior instance — the original thread on this same issue (alert 274) shows as auto-resolved above for that reason, not because it was addressed differently. Re-verified against this alert's own SARIF: same rule, same `isSoloHelpRequest` source, same `filterToolsForMcpScope(...)` sink (now line 1125, still byte-identical to `master`). Full trace and false-positive analysis is on the resolved thread just above this one — nothing new to add; expect this to keep re-appearing under new numbers on further pushes until either dismissed or the code is restructured.
